File: backend/comment-service/src/main/java/com/facebook/comment/service/CommentService.java
package com.facebook.comment.service;

import com.facebook.comment.payload.CommentDTO;
import com.facebook.comment.payload.request.CommentRequest;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.util.List;

public interface CommentService {
    List<CommentDTO> getCommentsByPostId(Long postId);
    
    Page<CommentDTO> getCommentsByPostId(Long postId, Pageable pageable);
    
    CommentDTO getCommentById(Long id);
    
    CommentDTO createComment(Long authorId, CommentRequest commentRequest);
    
    CommentDTO updateComment(Long id, Long userId, CommentRequest commentRequest);
    
    void deleteComment(Long id, Long userId);
    
    List<CommentDTO> getCommentsByAuthorId(Long authorId);
    
    CommentDTO voteComment(Long id, Long userId, boolean upvote);
    
    void updateCommentVoteCount(Long commentId, int voteCount);

    List<CommentDTO> getRepliesByParentId(Long parentId);

    long countByPostIdAndParentIdIsNull(Long postId);
} 

File: frontend/src/app/comment/services/comment.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../../environments/environment';
import { Comment, CommentRequest } from '../models/comment.model';

const API_URL = environment.commentApiUrl;

@Injectable({
  providedIn: 'root'
})
export class CommentService {
  constructor(private http: HttpClient) { }

  getCommentsByPostId(postId: number): Observable<Comment[]> {
    return this.http.get<Comment[]>(`${API_URL}/comments/post/${postId}`);
  }

  getPagedCommentsByPostId(postId: number, page: number = 0, size: number = 10): Observable<any> {
    return this.http.get(`${API_URL}/comments/post/${postId}/paged?page=${page}&size=${size}`);
  }

  getCommentById(id: number): Observable<Comment> {
    return this.http.get<Comment>(`${API_URL}/comments/${id}`);
  }

  createComment(commentRequest: CommentRequest): Observable<Comment> {
    return this.http.post<Comment>(`${API_URL}/comments`, commentRequest);
  }

  updateComment(id: number, commentRequest: CommentRequest): Observable<Comment> {
    return this.http.put<Comment>(`${API_URL}/comments/${id}`, commentRequest);
  }

  deleteComment(id: number): Observable<any> {
    return this.http.delete(`${API_URL}/comments/${id}`);
  }

  getCommentsByAuthorId(authorId: number): Observable<Comment[]> {
    return this.http.get<Comment[]>(`${API_URL}/comments/author/${authorId}`);
  }

  voteComment(id: number, upvote: boolean): Observable<Comment> {
    return this.http.post<Comment>(`${API_URL}/comments/${id}/vote?upvote=${upvote}`, {});
  }

  getReplies(commentId: number): Observable<Comment[]> {
    return this.http.get<Comment[]>(`${API_URL}/comments/${commentId}/replies`);
  }
} 